src/Plugin/Action/StateChange.php

```diff
@@ -194,7 +204,14 @@ public function calculateDependencies() {
    * {@inheritdoc}
    */
   public function execute(ContentEntityInterface $entity = NULL) {
-    $entity = $this->loadLatestRevision($entity);
+    // Create a new revision if entity is revisionable to fix some edge cases
+    // with performing moderation states which publishes the entity.
+    if ($entity->getEntityType()->isRevisionable()) {
+      $entity = $this->loadLatestRevision($entity);
+      $entity = $this->entityTypeManager->getStorage($entity->getEntityTypeId())
+       ->createRevision($entity, $entity->isDefaultRevision());
+    }
+
```

Comment:
yep as I said in the meeting, my concern here is duplicate revisions for other entities that use this module. If you don't find any issues with them. I'm fine with this fix.

Reply:
you may want to also set the meta data for the revision if you are creating a new revision here.
from Drupal\content_moderation\Form\EntityModerationForm
```
    if ($entity instanceof RevisionLogInterface) {
      $entity
        ->setRevisionCreationTime($this->time
        ->getRequestTime());
      $entity
        ->setRevisionLogMessage($form_state
        ->getValue('revision_log'));
      $entity
        ->setRevisionUserId($this
        ->currentUser()
        ->id());
    }
```